src/Build/BackEnd/Components/BuildRequestEngine/BuildRequestEngine.cs

```diff
@@ -1127,7 +1127,7 @@ private void IssueBuildRequests(BuildRequestEntry issuingEntry, FullyQualifiedBu
                         // we receive the configuration response before we enter the wait state.
                         newRequest = new BuildRequest(issuingEntry.Request.SubmissionId, GetNextBuildRequestId(),
                             request.Config.ConfigurationId, request.Targets, issuingEntry.Request.HostServices,
-                            issuingEntry.Request.BuildEventContext, issuingEntry.Request,
+                            issuingEntry.Request.CurrentMSBuildTask ?? issuingEntry.Request.BuildEventContext, issuingEntry.Request,
```

Comment:
My own concern is about this code, because it creates a regression in behavior which effect I cannot evaluate, because I do not know the code. After my change the `ChildProject.ParentProjectBuildEventContext <> ChildProject.ParentProject.BuildEventContext`. This may matter if there are maps that use BuildEventContext objects as keys. But I did not want to change the equality behavior of `BuildEventContext`, because this could mean an even greater regression. Somebody who knows the code very well should weigh in.

Reply:
Unfortunately, there's no such person available. I share your concern and will try to think about the implications, but I don't know if I'm going to be able to convince myself that it's "definitely ok".

Reply:
OK, how about a zero regression change where we add `TaskId2` to `BuildEventContext`? Its name reflects the current state of affairs - no one knows this area of code good enough to change it - ugly situation. So, we add to it. Writable field that does not participate in any Equals. It will be set to the MSBuild task Id in the project level `BuildEventContext`, so it is exposed in downstream projects as `ParentProjectBuildEventContext.TaskId2`.

Something has to be done. Correlating events is a nightmare.

Reply:
I think this should be safe - the `BuildRequestId` is just returning `GetHashCode`:
https://github.com/microsoft/msbuild/blob/7bcb2f961093e87277439420d351a9f3f3a8cdb8/src/Framework/BuildEventContext.cs#L172-L175

If someone is taking a dependency on this I'd be concerned.

Reply:
Would someone mind elaborating on why this is quite so risky? What I see is that either it does what was previously done or it uses something that's more accurate (prevented from messing up by turning it to null in the finally). I doubt someone would be too mad about us fixing a bug.